Modernise setup script to use uv consistently

Crooked-Krokr opened the pull request:
# Modernise setup script to use uv consistently

## Summary
Modernise setup script to use uv for faster, more reliable dependency management

## Changes
- Add upfront check for uv availability with installation guidance
- Replace mixed pip/uv usage with pure uv throughout
- Addition of pre-commit missing warning
- Improve status reporting to show both Python and uv versions
- Add reminder about venv activation

## Benefits
- **Faster installs**: uv is significantly quicker than pip for large dependency trees
- **Better reliability**: uv handles dependency conflicts more intelligently  
- **Improved developer experience**: Clear error messages and automatic tooling setup
- **Modern tooling**: Aligns with current Python ecosystem best practices

## Testing
- Tested on Ubuntu 22.04 with WSL
- Verified behaviour with and without existing venv
- Confirmed all dependencies install correctly
- pre-commit passed across the board of checks

```bash
(.venv) crooks@LAPTOP-QJ718NJD:/mnt/x/git/server-1$ pre-commit run --all-files
🐶 Ruff Linter...........................................................Passed
🐶 Ruff Formatter........................................................Passed
🐍 Check Python AST......................................................Passed
🔠 Check for case conflicts..............................................Passed
ℹ️  Check docstring is first.............................................Passed
🧐 Check that executables have shebangs..................................Passed
｛ Check JSON files......................................................Passed
💥 Check for merge conflicts.............................................Passed
🔗 Check for broken symlinks.........................(no files to check)Skipped
✅ Check TOML files......................................................Passed
✅ Check code for common misspellings....................................Passed
🕵️  Detect Private Keys..................................................Passed
⮐  Fix End of Files......................................................Passed
🛑 Don't commit to stable branch.........................................Passed
✄  Trim Trailing Whitespace..............................................Passed
mypy.....................................................................Passed
gen_requirements_all.....................................................Passed
```

Crooked-Krokr committed:
Improve setup.sh: auto-detect required Python version and select newest matching interpreter

Crooked-Krokr committed:
setup.sh: robust Python version detection and venv management

- Parse requires-python from pyproject.toml and select the newest matching interpreter.
- Validate existing virtual environments against version spec.
- Add backup and auto-recreate logic for outdated venvs.
- Improve messaging for interpreter selection and venv handling.
- Add dry-run and skip-install hooks for CI/testing.
- update gitignore to ignore backed-up venv folders.

OzGav commented:
Simple lint issue to fix

marcelveldt commented:
While I appreciate the suggestion, I think this is a massive complication of the setup script for a simple issue.
Would it make sense to look into uv as that has native support to specify a python version for the environment and will handle all of this out of the box.

Crooked-Krokr commented:
> While I appreciate the suggestion, I think this is a massive complication of the setup script for a simple issue. Would it make sense to look into uv as that has native support to specify a python version for the environment and will handle all of this out of the box.

@marcelveldt You're absolutely right about uv being the more elegant solution. I appreciate the feedback and completely agree that my approach added unnecessary complexity.
I'd be happy to rework this using uv's native Python version management, which would be much cleaner and more maintainable. Alternatively, if documentation updates would be more valuable to the project, I could focus there instead.

What would be most useful for the codebase?

marcelveldt commented:
Well, this PR in it's current shape cant be accepted. Its like shooting a mosquito with a rocket.
If the issue about the wrong versioned python venv is real, we could also consider starting by a warning in the docs and/or setup script log.

But a more elegant way to handle this nowadays is to let uv deal with setting up the env as it can do it all at once.

Crooked-Krokr committed:
Reverted the helper functions for checking python versions against the pyproject.toml.
This is in an effort to simplify the original PR in order to make it easier to maintain.

Crooked-Krokr committed:
reverted change to settings.json to move the interpreter declaration into user settings. This will ensure the repository settings.json is unnecessarily changed.

Crooked-Krokr committed:
updated to use uv to create the venv as per pyproject.toml requirements.

Crooked-Krokr committed:
Updated to carry any named venv folder through the script to be activated when running this file. Previously, .venv was getting activated regardless of the passed env_name argument.

Crooked-Krokr committed:
removed a message about activating the source because we've already done that by this stage via "source "$env_name/bin/activate"" above.

Crooked-Krokr committed:
Updated pre-commit to warn early if missing. If someone has made their own venv from another method thant setup.sh then it could be worth flagging.

Crooked-Krokr committed:
Adding a check for uv in preparation for changing the pip stuff out for uv.

Crooked-Krokr committed:
consolidating project dependency management to uv instead of relying on both pip and uv.

Crooked-Krokr committed:
Dont need the test file now that the setup script is far simpler than before.

Crooked-Krokr committed:
Added some helping console output about the current python and uv version.
Also added how to activate the venv as a nudge to do it because although activated in the script, it'll need activated again.

Crooked-Krokr committed:
Merge branch 'dev' into fix/setup-python-detection

Crooked-Krokr commented:
> But a more elegant way to handle this nowadays is to let uv deal with setting up the venv as it can do it all at once.

I see that now, I can't believe how much smoother that was with uv; I couldn't work out how I'd missed it until reading its only 8 months old...  😅
Changes are committed but I've kicked this PR back to draft so I can update the description above tomorrow for review.

Crooked-Krokr committed:
Updated to a clearer. less verbose warning on missing pre-commit.

marcelveldt commented on scripts/setup.sh:
```diff
@@ -1,28 +1,41 @@
+if command -v pre-commit &>/dev/null; then
+  pre-commit install
+else
+  echo "⚠️  pre-commit not available. Install with: uv pip install pre-commit"
```
pre-commit will be installed with the test requirements isnt it ?

Crooked-Krokr replied:
My concern was if the dependency or project file was changed in future; this would flag it. Likely not required though I agree as the pre-commit install would fail regardless.
I can remove this back to just pre-commit install if you would prefer?

marcelveldt commented:
This looks much cleaner now indeed.

Crooked-Krokr committed:
Merge branch 'dev' into fix/setup-python-detection

Crooked-Krokr commented:
> This looks much cleaner now indeed.

Thanks, I think the pythonic rust is flaking off now. 😅

Crooked-Krokr committed:
Merge branch 'music-assistant:dev' into fix/setup-python-detection

marcelveldt reviewed (approved):
Thanks, @Crooked-Krokr !
